### DISCUSSION TITLE
Textual in WASM?

### DISCUSSION THREAD

**Discussion by showipintbri** [Textualize/textual#2764] [Ideas]
Will & Team,
What is the current level of Textual functionality in WASM Python? (pyscript or pyodide)

I'd love to be able to use Textual to build 1 single interface for both terminal and web based environments instead of building 1 for terminal interaction and 1 for web based interaction.

Is anyone already using Textual in WASM?

**Comment by davep** [comment]
Someone was [experimenting with pyscript late on last year](https://discord.com/channels/1026214085173461072/1033752599112994867/1044732418868912170), and someone [posted last month that they'd got their app running in pyodide](https://discord.com/channels/1026214085173461072/1033752599112994867/1103409153474560171).

**Comment by pmp-p** [comment]
Hi, I have only found a few things that would prevent Textual from running on a python wasm runtime "out of the box"

First is that textual.worker.Worker will not enforce thread = False even though 
`( not hasattr(sys, "_emscripten_info") or sys._emscripten_info.pthreads )` 
also if threading is not available and worker callable is not async, the WorkerError("Unsupported attempt to run a thread worker") should be turned into a warning/console.warn and attempt sync execution.

The annoying two `print` in texual/app.py when using pilot mode ( easiest way i've found to start the textual app as a normal async task without any tty check ) but that's only a problem if select does not work properly with stdin ( pyodide maybe ?).

It may also have some trouble when app code is coming from stdin and [not catching OSError here](https://github.com/Textualize/textual/blob/e819e97c2984fd2e75df9b06dcf88c135fca26a9/src/textual/_path.py#L70)

The monkey patch I used to run [codebrowser.py](https://github.com/Textualize/textual/blob/main/examples/code_browser.py) on pygbag

```py
import asyncio
import select
import os

import textual.events
import textual.driver
import textual.drivers.headless_driver
import textual._xterm_parser

class HeadlessDriver(textual.driver.Driver):
    def send_event(self, event: textual.events.Event) -> None:
        self._loop.create_task(self._app._post_message(event))

    def _get_terminal_size(self) -> tuple[int, int]:
        return tuple(map(int, __import__("os").get_terminal_size()))

    def write(self, data: str) -> None:
        sys.__stdout__.write(data)

    def start_application_mode(self) -> None:
        loop = asyncio.get_running_loop()
        loop.create_task(self.input_handler())

    def disable_input(self) -> None:
        ...

    def stop_application_mode(self) -> None:
        ...

    async def input_handler(self):
        import sys, os
        from platform import window
        import textual._xterm_parser

        STDIN = sys.stdin.fileno()

        def more_data() -> bool:
            return select.select([STDIN], [], [], 0)[0]

        parser = textual._xterm_parser.XTermParser(more_data, debug=self._debug)
        while True:
            await asyncio.sleep(0)
            if select.select([STDIN], [], [], 0)[0]:
                try:
                    payload = os.read(STDIN, 1024)
                    if payload:
                        for event in parser.feed(payload.decode('utf-8')):
                            self.process_event(event)
                except OSError:
                    continue


textual.drivers.headless_driver.HeadlessDriver = HeadlessDriver
```

There is surely a better way to integrate python wasm but i only spent an hour on this.

![shot-2023-11-16_1700151783](https://github.com/Textualize/textual/assets/16009100/e8d1726c-5c30-4c92-b984-891882de7a04)

probably some font width / xterm.js specific problems here with textual-paint
![shot-2023-11-16_1700163808](https://github.com/Textualize/textual/assets/16009100/f6c691ab-e5c4-4738-bd3c-c9375acba3d2)

**Comment by mbund** [comment]
Does anyone have a fully working repo or single file of this? I am interested in being able to run textual in WASM in wasmtime, but I can't seem to adapt the approach @pmp-p describes.

To compile it at all with [componentize-py](https://github.com/bytecodealliance/componentize-py) I need to add:

```
import os

if not hasattr(os, "getuid"):
    os.getuid = lambda: 0
```

The next issues is

```
Traceback (most recent call last):
  File "/0/app.py", line 72, in run
  File "/1/textual/app.py", line 612, in __init__
  File "/1/textual/app.py", line 1508, in get_driver_class
ModuleNotFoundError: No module named 'textual.drivers.linux_driver'
```

So I set

```
textual.constants.DRIVER = "textual.drivers.headless_driver:HeadlessDriver"
```

But then, I get an error on `asyncio.run`:

```
Traceback (most recent call last):
  File "/0/app.py", line 73, in run
  File "/1/textual/app.py", line 2251, in run
  File "/python/asyncio/runners.py", line 203, in run
  File "/python/asyncio/runners.py", line 59, in __enter__
  File "/python/asyncio/runners.py", line 146, in _lazy_init
  File "/python/asyncio/events.py", line 844, in new_event_loop
  File "/python/asyncio/events.py", line 732, in new_event_loop
  File "/python/asyncio/unix_events.py", line 61, in __init__
  File "/python/asyncio/selector_events.py", line 66, in __init__
  File "/python/asyncio/selector_events.py", line 120, in _make_self_pipe
  File "/python/socket.py", line 621, in _fallback_socketpair
PermissionError: [Errno 2] Permission denied
```

Applying the HeadlessDriver monkey patch doesn't seem to help. Has successfully been able to run Textual in WASM with WASI on the CLI? Or have a suggestion on how it can be done? I am pretty new to Textual as a library and would appreciate some guidance.

I am on v7.5.0 of Textual. [app.py:2251](https://github.com/Textualize/textual/blob/v7.5.0/src/textual/app.py#L2251) is where it threw on `asyncio.run`. I am also open to not use componentize-py, if its [limitations](https://github.com/bytecodealliance/componentize-py?tab=readme-ov-file#known-limitations) are the problem. Wasip1 is good enough for my use case.

Here is my debugging workflow. I cloned the [componentize-py](https://github.com/bytecodealliance/componentize-py) repo and am in the examples directory (just so I can get the wit files easily). I then edit `app.py` and run the following:
```
$ componentize-py -d ../../wit -w wasi:cli/command@0.2.0 componentize app -o cli.wasm
$ wasmtime run cli.wasm
```
